jetstream_pt/cache_manager.py

```diff
@@ -225,3 +225,61 @@ def update(self, xk, xv):
       self.k_scaler[batch, :, self.input_pos, :] = kscale.squeeze(2)
       self.v_scaler[batch, :, self.input_pos, :] = vscale.squeeze(2)
     return self.cache_k, self.cache_v, self.k_scaler, self.v_scaler
+
+
+class PageKVCacheGenerate:
+  """Page attention kvache generator without quantization"""
+
+  def __init__(
+      self,
+      cache_k: torch.Tensor,  # previous cache
+      cache_v: torch.Tensor,  # previous cache
+      page_token_indices: jax.Array,  # page and token indices for the cache
+      sharding,
+      env=None,
+  ):
+    super().__init__()
+    self.cache_k = cache_k
+    self.cache_v = cache_v
+    self.page_token_indices = page_token_indices
+    self.sharding = sharding
+    self.env = env
+
+  def update(self, key, value):
+    """Update kv cache"""
+    keyj, valuej = torchjax.to_torch((key, value))
+
+    def _update(cache, x):
+      x = x.squeeze(2).transpose(1, 0).jax()
+      x = x[:, self.page_token_indices[2], :]
+      head, _, page_size, dim = cache.shape
+      selected_cache = cache[:, self.page_token_indices[0], :, :]
+      selected_cache = selected_cache.reshape((head, -1, dim))
+
+      selected_cache = selected_cache.at[:, self.page_token_indices[1], :].set(
+          x
+      )
+      selected_cache = selected_cache.reshape((head, -1, page_size, dim))
+
+      cache = cache.at[:, self.page_token_indices[0], :, :].set(selected_cache)
+      return cache
+
+    # pylint: disable-next=all
+    self.cache_k._elem = _update(self.cache_k._elem, keyj)
```

Comment:
self.cache_k is torch.tensor; so you can operate on it directly without ._elem

Reply:
Thanks Han! Do you mean doing torch matrix update instead of jax matrix update?